lib/dockerregistry/checks.go

```diff
@@ -245,3 +253,195 @@ func (check *ImageSizeCheck) Run() error {
 
 	return nil
 }
+
+// MKImageVulnCheck returns an instance of ImageVulnCheck which
+// checks against images that have known vulnerabilities.
+// nolint[funlen]
+func MKImageVulnCheck(
+	syncContext SyncContext,
+	newPullEdges map[PromotionEdge]interface{},
+	severityThreshold int,
+	customVulnProducer ImageVulnProducer,
+) *ImageVulnCheck {
+	return &ImageVulnCheck{
+		syncContext,
+		newPullEdges,
+		severityThreshold,
+		customVulnProducer,
+	}
+}
+
+// Run is a function of ImageVulnCheck and checks that none of the
+// images to be promoted have any severe vulnerabilities.
+// nolint[errcheck]
+func (check *ImageVulnCheck) Run() error {
+	var populateRequests PopulateRequests = func(
+		sc *SyncContext,
+		reqs chan<- stream.ExternalRequest,
+		wg *sync.WaitGroup) {
+		srcImages := make(map[PromotionEdge]interface{})
+		for edge := range check.PullEdges {
+			srcImage := PromotionEdge{
+				SrcRegistry: edge.SrcRegistry,
+				SrcImageTag: edge.SrcImageTag,
+				Digest:      edge.Digest,
+			}
+			// Only check the vulnerability for the source image if it
```

Comment:
You should only be checking for the digest field. There can be multiple sources for the same digest (aka "redundant" promotions as discussed in your design doc IIRC).

Reply:
done